daft/dataframe/dataframe.py

```diff
@@ -729,6 +729,69 @@ def write_deltalake(
 
         return with_operations
 
+    @DataframePublicAPI
+    def write_lance(
+        self,
+        uri: Union[str, pathlib.Path],
+        mode: str = "create",
+        io_config: Optional[IOConfig] = None,
+        **kwargs: Any,
+    ) -> "DataFrame":
+        """
+        Writes the DataFrame to a Lance table
+        Args:
+          uri: The URI of the Lance table to write to
+          io_config (IOConfig, optional): configurations to use when interacting with remote storage.
+          mode (str, optional): Operation mode of the write. Defaults to "create". See `lance.write_dataset` for more information.
+          **kwargs: Additional keyword arguments to pass to the Lance writer. See `lance.write_dataset` for more information.
+        Example:
+        --------
+
+
+        >>> df = daft.from_pydict({"a": [1, 2, 3, 4]})
+        >>> df.write_lance("/tmp/lance/my_table.lance")
+        >>> daft.read_lance("/tmp/my_table.lance")
+
+
+        # Pass additional keyword arguments to the Lance writer
+        # All additional keyword arguments are passed to `lance.write_dataset`
+        >>> df.write_lance("/tmp/lance/my_table.lance", mode="overwrite", max_bytes_per_file=1024)
+
+        """
+        from daft import from_pydict
+        from daft.io.object_store_options import io_config_to_storage_options
+
+        try:
+            import lance
+            import pyarrow as pa
+
+        except ImportError:
+            raise ImportError("lance is not installed. Please install lance using `pip install getdaft[lance]`")
+        tbl = self.to_arrow()
```

Comment:
How hard do you think it would be performing a per-partition-write similar to what we do for [iceberg](https://github.com/Eventual-Inc/Daft/blob/b7e6e41e627035fe0a5183b38baf35560befdac2/daft/execution/execution_step.py#L374) and deltalake?

That way we can distributed the write rather having all the data come back to the head node.

Reply:
🤔 Hmm, I'll take a look into it.

Reply:
@samster25 I was looking at this a bit closer, and it doesn't seem that lance AFAICT has a way to easily support partitioned writes. 

there is a `fragment.write` method, but that only writes the data fragments and we'd have to manually have to write all of the additional metadata `_transactions, _versions, _latest.manifest`

Reply:
@wjones127 or @westonpace, do either of you know if `lance` supports this?

Reply:
We do support that. For example, to perform an append:

1. To write the data files, use the [write_fragments()](https://lancedb.github.io/lance/api/python/lance.html#lance.fragment.write_fragments) method. This can be done on workers for each partition or CPU.
2. Then use [Operation.Append()](https://lancedb.github.io/lance/api/python/lance.html#lance.dataset.LanceOperation.Append) to create an operation to be committed. This takes the fragment metadata output by `write_fragments()` as it's input.
3. Pass the operation to [Dataset.commit()](https://lancedb.github.io/lance/api/python/lance.html#lance.LanceDataset.commit) to write all the meta files and commit the operation.

There are various examples in the API reference for other operations, such as overwrite, delete, and merge.